Colour injection

elduddz opened the pull request:
- adding notation formatting on logging for output to screen
- Logging measures on how many types of change are occurring
- Added an Import-RequiredModule build script and associated function to pipeline
- added notation for errors
- increase on version to 5.1

** need to add section to readme about testing locally and getting required modules**

ld-at-work committed:
injecting some colour and format to the start of the run

elduddz committed:
updated pester to 4.6

elduddz committed:
adding some more formatting to the output written

ld-at-work committed:
adding Import-RequiredModules to build pipeline script

ld-at-work committed:
Merge branch 'colourInjection' of https://github.com/UKHO/ADM-PSmodule into colourInjection

ld-at-work committed:
moving module install step above run

ld-at-work committed:
adding Install-AcModule function

ld-at-work committed:
error state added to notation

ld-at-work committed:
all Write host in object processes are now write-color

ld-at-work committed:
didnt clock that purple wasnt a choice

ld-at-work committed:
upped version to cover what I think is a major change

ld-at-work committed:
trying to reduce the text in the header a bit

elduddz committed:
there is already a property that finds the CN for us `$this.Name`

elduddz committed:
error format to 1 tab

elduddz committed:
Using existing properties of Name and Domain.FQDN to get shorter values

elduddz committed:
return measures of change

incremental counts on Create Add and Remove on various objects

elduddz commented:
I want to make sure to squash this on completing.

elduddz committed:
replaced to use UPN rather than SamAccountName

ld-at-work committed:
some formatting around missing users

elduddz commented:
Im really happy:
![image](https://user-images.githubusercontent.com/338863/51744427-bb895180-2097-11e9-8e79-de986a43a14c.png)

elduddz commented:
Im really happy with the output, I did upload a picture of a test against a small section, but realised it gave away too much :)

ld-at-work committed:
logging of group name set to groupMember not group

ld-at-work committed:
more process logging

ld-at-work committed:
reverting version to 4.3.1

I rememebr this being the issue of why it stopped working last time?

ld-at-work committed:
removed pointer completely

MarcDenman commented on UKHO.ADM/Functions/Update-AD.ps1:
```diff
@@ -6,21 +6,44 @@ function Update-AD {
     process {
+        $fg = "White"
+
+        Write-Color "A change plan has been geneated and is show below.`n",
```
Lose this first line and just start with `Actions are indicated with the following symbols`

elduddz replied:
missed this one off, done now

MarcDenman commented on UKHO.ADM.Tests/functions/Update-AD.Tests.ps1:
```diff
@@ -2,7 +2,7 @@ InModuleScope $mut {
         Context "with valid cd" {
             Mock Generate-ConfigurationObject       
-            Mock Get-ADChanges
+            Mock Get-ADChanges { return @{ "CreatedOUs" = 1}}
```
This seems.... odd. Is this for test that didn't get written? Does it fix a test or something?

elduddz replied:
it shows that the CreatedOUs property exists, so returned to screen `Created OUs: 1`

elduddz replied:
the Plan info returned is blank on the values otherwise

MarcDenman replied:
But you aren't asserting on anything? It is just to display to the screen during the test?

elduddz replied:
it was a noddy thing for me to visually see that i was returning the property to screen.

MarcDenman commented on UKHO.ADM/Functions/Update-AD.ps1:
```diff
@@ -6,21 +6,44 @@ function Update-AD {
+        $fg = "White"
+
+        Write-Color "A change plan has been geneated and is show below.`n",
+        "Actions are indicated with the following symbols:`n",
```
This Write-Color is hard to read/comprehend. Any chance it can be made easier?

Maybe lose the $fg variable and just write "white"? Or rename $fg, whilst it is the foreground colour, it takes a couple seconds to realise that actually this is the "default" text colour.

Would having multiple `Write-Color` help?

```powershell
Write-Color "Actions are indicated with the following symbols:" -Color White
Write-Color "`t+"," adding or creating" -Color Green, White
Write-Color "`t-"," removing or deleting",-Color Red, White
```

elduddz replied:
probably easier to read, so might do that, I do do that in other places

MarcDenman commented on UKHO.ADM/Functions/Update-AD.ps1:
```diff
@@ -6,21 +6,44 @@ function Update-AD {
+        "`tx", " errors have occurred`n" `
+        -Color $fg, 
+        $fg,
+        Green,$fg,
```
Global variables are bad but maybe in this case it could be helpful. Would it make sense to have $AddCreateColor = "Green" defined globally and then use that instead. Not sure how often it will be used but might help. Would need to be done for all global colours.

elduddz replied:
I think I am happy that the only place we are using these colours are in the objects, so might be some value in setting the value once and just update this file. will leave as is for now.

MarcDenman replied:
I will open this as a separate issue.

MarcDenman commented on UKHO.ADM/Functions/Update-AD.ps1:
```diff
@@ -6,21 +6,44 @@ function Update-AD {
 
         $ADChanges = Get-ADChanges($out)
 
+        Write-Color "Plan:`n",
```
Not really a plan :p `Change Summary` instead?

MarcDenman commented on UKHO.ADM/Objects/Objects.ps1:
```diff
@@ -131,25 +137,34 @@ class ADChanges {
                 try {
                     New-ADOrganizationalUnit $ADOrganisationalUnit.Name -Path $ADOrganisationalUnit.ParentOrganistaionalUnit -Server $ADOrganisationalUnit.Domain.DomainController -Credential $ADOrganisationalUnit.Domain.Credential
-                    Write-Host "CREATED OU $($ADOrganisationalUnit.DistinguishedName)"
+                    Write-Colour -LinesBefore 2 "`t+ CREATED OU $($ADOrganisationalUnit.Name) on $($ADOrganisationalUnit.Domain.FQDN)" -Color Green
```
`-LinesBefore 2` isn't intuitive, I had to read the docs to work out it added 2 blank lines. My first thought was that it somehow coloured the two lines before it. Is there a better way to document/implement this?

elduddz replied:
I think there might be an overhead to calling Write-Color  the alternative would be 
```
`n`n
```

MarcDenman replied:
OK. Leave as is, neither are particularly great options.  I do think it is unexpected behaviour that `Write-Color` also inserts blank lines. Wouldn't be what I expect it to do.

MarcDenman commented on UKHO.ADM/Objects/Objects.ps1:
```diff
@@ -237,27 +267,34 @@ class ADChanges {
 
     RemoveGroupMemberFromG($groupMember, [ADGroup] $group) {
-        $outputString = "REMOVE GROUP $($groupMember.DistinguishedName) FROM GROUP $($group.DistinguishedName)"
+        $outputString = "`t`tREMOVE GROUP $($groupMember.DistinguishedName) FROM GROUP $($group.DistinguishedName)"
```
Unnecessary modification, this output string isn't used to display anywhere

elduddz replied:
ah whoops, I meant to leave that alone

MarcDenman commented on UKHO.ADM/Objects/Objects.ps1:
```diff
@@ -237,27 +267,34 @@ class ADChanges {
                 $this.StringContent.Add($outputString, $true)
-    
-                Write-Host $outputString -ForegroundColor Red
+                Write-Color "x Failed to add user $($user.UPN) to group $($group.Name) on $($group.Domain.DistinguishedName):" -Color Magenta                        
```
Excessive whitespace at the end

elduddz replied:
odd, that wasnt picked up by the PSAnalyser

MarcDenman commented on UKHO.ADM/UKHO.ADM.psd1:
```diff
@@ -12,7 +12,7 @@
     
     # Version number of this module.
-    ModuleVersion = '5.0.0'
+    ModuleVersion = '6.0.0'
```
I don't know if this is a major increment. Is there any breaking changes around the API?

elduddz replied:
we are referencing our first required module but the call to the start is the same, so maybe 5.1.0

MarcDenman commented on Build/Install-AcModule.ps1:
```diff
@@ -0,0 +1,32 @@
+function Install-AcModule 
```
What does Ac stand for?

elduddz replied:
I cannot remember this is a function copied from the BuildAndDeploy module.

maybe it means avoid clobber?

MarcDenman replied:
I don't think this entire function needs to exist at all, what is wrong with `Install-Module -name -requiredVersion`

MarcDenman replied:
But, it isn't doing much harm and maybe someone knows something I don't. The inclusion of `-SkipPublisherCheck` is a tad worrying though.

elduddz replied:
Let's raise an issue to investigate

Commit:
remove use of

Commit:
one use of Write-Color per line

Commit:
Summary not Plan

Commit:
removing 		 from outstring

Commit:
excessive space removed

Commit:
reversioned to 5.1

Change reviewed as changes to internal function
although there is a required module the overall call hasnt changed.

Commit:
some more changes to wording around actions and summary

Commit:
removed space between parameter and declaration of parameter
